extension/android/jni/jni_layer_asr.cpp

```diff
@@ -0,0 +1,175 @@
+      }
+      uint8_t next_byte = static_cast<uint8_t>(str[i + 1]);
+      if ((byte & 0xE0) == 0xC0 &&
+          (next_byte & 0xC0) == 0x80) { // 2-byte sequence
+        i += 1;
+      } else if (
+          (byte & 0xF0) == 0xE0 && (next_byte & 0xC0) == 0x80 &&
+          (i + 2 < length) &&
+          (static_cast<uint8_t>(str[i + 2]) & 0xC0) ==
+              0x80) { // 3-byte sequence
+        i += 2;
+      } else if (
+          (byte & 0xF8) == 0xF0 && (next_byte & 0xC0) == 0x80 &&
+          (i + 2 < length) &&
+          (static_cast<uint8_t>(str[i + 2]) & 0xC0) == 0x80 &&
+          (i + 3 < length) &&
+          (static_cast<uint8_t>(str[i + 3]) & 0xC0) ==
+              0x80) { // 4-byte sequence
+        i += 3;
+      } else {
+        return false; // Invalid sequence
+      }
+    }
+  }
+  return true; // All bytes were valid
+}
+
+std::string token_buffer;
+} // namespace
+
+namespace executorch_jni {
+
+class ExecuTorchASRCallbackJni
+    : public facebook::jni::JavaClass<ExecuTorchASRCallbackJni> {
+ public:
+  constexpr static const char* kJavaDescriptor =
+      "Lorg/pytorch/executorch/extension/audio/ASRCallback;";
+
+  void onResult(std::string result) const {
+    static auto cls = ExecuTorchASRCallbackJni::javaClassStatic();
+    static const auto method =
+        cls->getMethod<void(facebook::jni::local_ref<jstring>)>("onResult");
+
+    token_buffer += result;
+    if (!utf8_check_validity(token_buffer.c_str(), token_buffer.size())) {
+      ET_LOG(
+          Info, "Current token buffer is not valid UTF-8. Waiting for more.");
+      return;
+    }
+    result = token_buffer;
+    token_buffer = "";
+    facebook::jni::local_ref<jstring> s = facebook::jni::make_jstring(result);
+    method(self(), s);
+  }
+};
+
+class ExecuTorchASRJni : public facebook::jni::HybridClass<ExecuTorchASRJni> {
+ private:
+  friend HybridBase;
+  std::unique_ptr<example::WhisperRunner> runner_;
+
+ public:
+  constexpr static auto kJavaDescriptor =
+      "Lorg/pytorch/executorch/extension/audio/ASRModule;";
+
+  static facebook::jni::local_ref<jhybriddata> initHybrid(
+      facebook::jni::alias_ref<jclass>,
+      facebook::jni::alias_ref<jstring> model_path,
+      facebook::jni::alias_ref<jstring> tokenizer_path) {
+    return makeCxxInstance(model_path, tokenizer_path);
+  }
+
+  ExecuTorchASRJni(
+      facebook::jni::alias_ref<jstring> model_path,
+      facebook::jni::alias_ref<jstring> tokenizer_path) {
+#if defined(ET_USE_THREADPOOL)
+    // Reserve 1 thread for the main thread.
+    int32_t num_performant_cores =
+        ::executorch::extension::cpuinfo::get_num_performant_cores() - 1;
+    if (num_performant_cores > 0) {
+      ET_LOG(Info, "Resetting threadpool to %d threads", num_performant_cores);
+      ::executorch::extension::threadpool::get_threadpool()
+          ->_unsafe_reset_threadpool(num_performant_cores);
+    }
+#endif
+
+    // create runner
+    runner_ = std::make_unique<example::WhisperRunner>(
```

Comment:
This should be an interface, like ASRRunner or something like that

Reply:
Does it mean this is not runnable without QNN?